### PR TITLE
Make sure we never access deleted objects in callbacks.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/resources/ble_esp32.cc [hunk 1]
```diff
@@ -82,15 +82,145 @@ class DiscoveredPeripheral : public DiscoveredPeripheralList::Element {
...
+class TokenResourceMap {
+ public:
+  ~TokenResourceMap() {
+    if (capacity != -1) {
+      free(entries);
+    }
+  }
+
+  // Returns false if there was a malloc error.
+  bool add(BleResource* resource, word* result) {
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:102]: Maybe change this to uword consistently? This makes the wrapping around behavior better.

File: src/resources/ble_esp32.cc [hunk 2]
```diff
@@ -82,15 +82,145 @@ class DiscoveredPeripheral : public DiscoveredPeripheralList::Element {
...
+        left = middle + 1;
+      } else {
+        right = middle - 1;
+      }
+    }
+    return -1;
+  }
+
+  bool resize(int new_capacity) {
+    auto new_entries = unvoid_cast<TokenResourceEntry*>(malloc(new_capacity * sizeof(TokenResourceEntry)));
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:190]: Can you use realloc instead?

File: src/resources/ble_esp32.cc [hunk 3]
```diff
@@ -82,15 +82,145 @@ class DiscoveredPeripheral : public DiscoveredPeripheralList::Element {
...
+      free(entries);
+      entries = null;
+      capacity = -1;
+      length = 0;
+    }
+  }
+
+ private:
+  static const int kInitialLength = 4;
+  static word sequence_counter;
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:170]: Why is this static?

File: src/resources/ble_esp32.cc [hunk 4]
```diff
@@ -653,8 +840,9 @@ class BleRemoteDeviceResource : public ServiceContainer<BleRemoteDeviceResource>
 
   int connect(uint8 own_addr_type, ble_addr_t* addr) {
     if (connected_) return BLE_ERR_SUCCESS;
+    // The 'connect' primitive ensured that the token is set.
     int err = ble_gap_connect(own_addr_type, addr, 3000, null,
-                              BleRemoteDeviceResource::on_event, this);
+                              BleRemoteDeviceResource::on_event, this->token());
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:845]:
```suggestion
                              BleRemoteDeviceResource::on_event, token());
```

File: src/resources/ble_esp32.cc [hunk 5]
```diff
@@ -1015,7 +1214,7 @@ BleServiceResource::_on_characteristic_discovered(const BleCallbackScope& scope,
                                 start_handle(),
                                 end_handle(),
                                 BleServiceResource::on_descriptor_discovered,
-                                this);
+                                this->token());
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:1217]:
```suggestion
                                token());
```

File: src/resources/ble_esp32.cc [hunk 6]
```diff
@@ -82,15 +82,145 @@ class DiscoveredPeripheral : public DiscoveredPeripheralList::Element {
...
+  bool add(BleResource* resource, word* result) {
+    if (!reserve_space()) return false;
+    word token = sequence_counter++;
+    *result = token;
+    TokenResourceEntry entry = {token, resource};
+    entries[length++] = entry;
+    return true;
+  }
+
+  bool reserve_space() {
```
Comment (kasperl) [inline src/resources/ble_esp32.cc:111]: I think you should avoid these super long methods inside the class definitions. You're asking the compiler to inline them.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/resources/ble_esp32.cc (hunk 1 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 3 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 4 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 5 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 6 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

File: src/resources/ble_esp32.cc (hunk 2 above)
Comment (floitsch) [reply src/resources/ble_esp32.cc]: done.

**Pull request merged**